bundles/org.openhab.binding.smhi/src/main/java/org/openhab/binding/smhi/internal/SmhiHandler.java

```diff
@@ -0,0 +1,421 @@
+
+/**
+ * The {@link SmhiHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Anders Alfredsson - Initial contribution
+ */
+@NonNullByDefault
+public class SmhiHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(SmhiHandler.class);
+
+    private @NonNullByDefault({}) SmhiConfiguration config;
+
+    private final HttpClient httpClient;
+    private @Nullable SmhiConnector connection;
+    private ZonedDateTime currentHour;
+    private ZonedDateTime currentDay;
+    private @Nullable TimeSeries cachedTimeSeries;
+    private boolean hasLatestForecast = false;
+    private @Nullable Future<?> forecastUpdater;
+    private @Nullable Future<?> instantUpdate;
+
+    public SmhiHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.httpClient = httpClient;
+        this.currentHour = calculateCurrentHour();
+        this.currentDay = calculateCurrentDay();
+    }
+
+    /**
+     * Handles commands sent to channels. Since all values are read-only, only REFRESH commands are allowed.
+     * Sending REFRESH to any item updates all items, since all values are returned in the response from Smhi.
+     * Therefore there's a wait of 5 seconds before the values are fetched, in which time all other commands are
+     * blocked, to prevent spamming Smhi's API.
+     *
+     * @param channelUID
+     * @param command
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            synchronized (this) {
+                Future<?> localRef = instantUpdate;
+                if (localRef == null || localRef.isDone()) {
+                    instantUpdate = scheduler.schedule(this::getUpdatedForecast, 5, TimeUnit.SECONDS);
+                } else {
+                    logger.debug("Already waiting for scheduled refresh");
+                }
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(SmhiConfiguration.class);
+        if (config == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+            return;
+        }
+        connection = new SmhiConnector(httpClient);
+
+        // Check which channel groups are selected in the config.
+        List<Channel> channels = new ArrayList<>();
+        channels.addAll(createChannels());
+        updateThing(editThing().withChannels(channels).build());
+
+        startPolling();
+        updateStatus(ThingStatus.ONLINE);
+        synchronized (this) {
+            Future<?> localRef = instantUpdate;
+            if (localRef == null || localRef.isDone()) {
+                instantUpdate = scheduler.schedule(this::getUpdatedForecast, 5, TimeUnit.SECONDS);
+            } else {
+                logger.debug("Already waiting for scheduled refresh");
+            }
+        }
```

Comment:
Please refactor this to its own method so that handleCommand and initialize can both use it.